### PR TITLE
Customizable/configurable status line

### PR DESCRIPTION
**Pull request by etienne-k**
### Summary

Building upon the feedback of #2420 and the discussion in #708, I refactored the status line implementation and made it configurable.

The configuration happens in the user's configuration file and distinguishes between three areas of the status line: **left**, **center** and **right**.

```
[ ... ... LEFT ... ... | ... ... ... ... CENTER ... ... ... ... | ... ... RIGHT ... ... ]
```

Within each area, status line elements can be defined. There are no particular restrictions regarding the order, and elements can occur multiple times or be removed completely.

The following elements can currently be configured (see `helix_view::editor::StatusLineElement`):

```
mode
spinner
file-name
file-encoding
file-type
diagnostics
selections
position
```

### Example

This would move the file name to the center, and add a file type indicated to the far right):

```toml
[editor.status-line]
left = ["mode", "spinner"]
center = ["file-name"]
right = ["diagnostics", "selections", "position", "file-encoding", "file-type"]
```

image

### Additional info

The default configuration keeps the same status line layout with the same elements as before. While I do see room for improvement here, I assume it'd better be done in another PR.

On top of the elements which were already present, I added the implementation for one additional element: the file type (disabled by default).

### CODE REVIEW

File: helix-term/src/ui/statusline.rs
Comment (archseer) [inline helix-term/src/ui/statusline.rs:23]: Is this extra struct really needed? You could map the config enum directly to simple functions that take a render context. Gutters work in a similar way but are a bit more complicated (a generator function that returns a closure since we repeatedly call it in a single render)

Comment (archseer) [reply helix-term/src/ui/statusline.rs:23]: The other advantage of exposing functions directly: you could allocate a single buffer for each section, then pass it into the render call and use `write!` instead of `format!`. That way you're not allocating a new string for each part.

i.e. for gutters

https://github.com/helix-editor/helix/blob/fcb8c50c9800338966b015863887f9f1c4f62659/helix-term/src/ui/editor.rs#L560-L561=

then each component just writes into it:

https://github.com/helix-editor/helix/blob/fcb8c50c9800338966b015863887f9f1c4f62659/helix-view/src/gutter.rs#L68=

**Review by archseer — COMMENTED** [review]
This is great! I apologize it took me a while to review.

File: helix-term/src/ui/statusline.rs
Comment (etienne-k) [reply helix-term/src/ui/statusline.rs]: Thanks again for the review, @archseer !

I finally found the time to have a look at it and got rid of `StatusLineElement` (see 5fe398d).
I took a more functional approach and simplified/generalized `get_render_function()` so it directly maps from an enum to a function. I like it much more like this; I hope that's more or less how you imagined it.

I didn't get rid of `format!()` in favor of `write!()` just yet, as it would require me obtaining an output stream directly instead of working with `Spans` buffers, which it turn would probably require refactoring beyond **statusline.rs**; I think that would become a bit too much for this PR.

File: helix-term/src/ui/statusline.rs
Comment (pickfire) [inline helix-term/src/ui/statusline.rs:175]:
```suggestion
                    _ if !visible => "   ",
```
I think we want this enum to be exhaustive so if we add a variant next time we won't miss this.

File: helix-term/src/ui/statusline.rs
Comment (pickfire) [inline helix-term/src/ui/statusline.rs:38]: I am a bit confused with the term `buffer` here, maybe `parts`?

Comment (etienne-k) [reply helix-term/src/ui/statusline.rs:38]: Yea, on the one hand, this is basically our rendering buffer. But at the same time, it's consisting of `Spans` (`spans`?).
If the name `parts` sounds more intuitive, I can ride with that as well.

File: helix-term/src/ui/statusline.rs
Comment (etienne-k) [reply helix-term/src/ui/statusline.rs:175]: Hmm, very good suggestion, but with the change, it seems like the compiler doesn't recognize the `match` arms as exhaustive anymore:

```rust
match context.doc.mode() {
    Mode::Insert if visible => "INS",
    Mode::Select if visible => "SEL",
    Mode::Normal if visible => "NOR",
    // If not focused, explicitly leave an empty space instead of returning None.
    _ if !visible => "   ",
}
```

```
error[E0004]: non-exhaustive patterns: `Normal`, `Select` and `Insert` not covered
   --> helix-term/src/ui/statusline.rs:166:23
    |
166 |                 match context.doc.mode() {
    |                       ^^^^^^^^^^^^^^^^^^ patterns `Normal`, `Select` and `Insert` not covered
    |
   ::: /home/user/Projects/helix/helix-view/src/document.rs:35:5
    |
35  |     Normal = 0,
    |     ------ not covered
36  |     Select = 1,
    |     ------ not covered
37  |     Insert = 2,
    |     ------ not covered
    |
    = help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
    = note: the matched value is of type `helix_view::document::Mode`
```

**Comment by mtoohey31** [comment]
Thanks for working on this, I just started using these changes and I'm loving the customizability already!

Not sure whether I should bring this up now, or wait until this is merged (or if this is even the right place for this discussion...), but I've written a "scope" element [here](https://github.com/mtoohey31/helix/tree/feat-statusline-scope) that uses the language server's `textDocument/documentSymbol` method to display the struct/function/etc. that the cursor is inside, inspired by [SmiteshP/nvim-navic](https://github.com/SmiteshP/nvim-navic). It's still a little rough; right now the icons are hard-coded to nerd-fonts icons, and I'm just triggering the request along with completion on idle, which might not be the right way to go about it, but cleaning things up shouldn't be too much work.

Is that something we'd be interested in supporting? And if so, should I hold on until after this is merged then create a separate PR, or is there another way I should go about things?

File: helix-term/src/ui/statusline.rs
Comment (mtoohey31) [inline helix-term/src/ui/statusline.rs:122]: With a small enough window (or a large sections) this overflows for me, resulting in a panic. I'm not sure how we want to prioritize the different sections when there isn't enough space for everything. I'm using `saturating_sub` as a quick fix in my own branch, but that doesn't look particularly good when the left section is larger than the right, as space gets wasted between the center and the right...

File: helix-term/src/ui/statusline.rs
Comment (sbromberger) [reply helix-term/src/ui/statusline.rs:171]: Sorry to just jump in, but couldn't you do something like
```
format!(
    " {} ",
    if visible {
        match {
            Mode::Insert => "INS",
            ...,
        } else {
            "    "
        }
    });
```
?
This way you have the variant checks, as well as pulling the visibility test outside the match.

**Comment by sudormrfbin** [comment]
@mtoohey31 I would recommend a separate PR that adds it as a new component; also we would want to avoid using nerd fonts by default (see https://github.com/helix-editor/helix/pull/2211).

File: helix-term/src/ui/statusline.rs
Comment (etienne-k) [reply helix-term/src/ui/statusline.rs]: I initially didn't want to clutter the code too much, but it doesn't look that bad after all, and it's a useful change. I committed this in 0f53b17d525e75bdc7c71e90cdb7578fc8fb38d4.

File: helix-term/src/ui/statusline.rs
Comment (etienne-k) [reply helix-term/src/ui/statusline.rs:122]: Thanks for the heads up! I'll look at this. A panic due to a certain configuration or due to a too small window must not happen.

I'm afraid we won't get around truncating elements, and I think the priority would be left > right > center (somehow feels more intuitive).

I'd also consider differentiating between a shortened version of elements (e.g. just a file name instead of a file path) when we detect there isn't enough space. But this would probably better fit in another PR.

Comment (etienne-k) [reply helix-term/src/ui/statusline.rs]: Alright, I added a `saturating_sub()` too: I can definitely see why the subtraction would cause an overflow.

Moreover, it's entirely correct for `center_max_width` to be `0` if there's not enough space left.

Changed in b2cc7f5.

**Comment by etienne-k** [comment]
> For the file type, could you make it configurable? I like seeing file type info but I don't want to take up multiple characters to show it. Having a config option that could override this default would allow me to use a glyph/icon without making assumptions about anyone else's font.

Hmm, _without_ introducing a dependency to devicons (or similar) in Helix, you would basically have to map filetypes to alternative labels (these can be single devicons) in your config.

It'd definitely be easy, but I think I'd first gather some extra input. E.g. maybe we'd be open to an opt-in (i.e. disabled by default) configuration option `devicons = "true"`. This would then allow us to map filetypes for you, and it would have the advantage of enabling devicons at many other places as well; not just in the statusline.

Either way, I believe it's worth opening a separate issue/discussion for this.

**Comment by sbromberger** [comment]
I'm currently blocked on #2862 (and, by extension, #2831). It'd be great to get this merged.

File: helix-term/src/ui/statusline.rs
Comment (sudormrfbin) [inline helix-term/src/ui/statusline.rs:231]: The contents of these functions can be inlined, I don't see the need for splitting them like this?

File: helix-view/src/editor.rs
Comment (sudormrfbin) [inline helix-view/src/editor.rs:144]: Let's name this `statusline`, similar to `bufferline` in https://github.com/helix-editor/helix/pull/2759.

**Comment by sudormrfbin** [comment]
Could you add section to the docs at https://github.com/helix-editor/helix/blob/master/book/src/configuration.md#editor ?

File: helix-view/src/editor.rs
Comment (etienne-k) [reply helix-view/src/editor.rs]: Corrected in f93ee65.

Comment (etienne-k) [reply helix-view/src/editor.rs]: Changed with c0a1870.

File: helix-term/src/ui/statusline.rs
Comment (etienne-k) [reply helix-term/src/ui/statusline.rs]: Changed with c0a1870.

File: helix-term/src/ui/statusline.rs
Comment (etienne-k) [reply helix-term/src/ui/statusline.rs]: Changed with c0a1870.

File: helix-term/src/ui/statusline.rs
Comment (etienne-k) [reply helix-term/src/ui/statusline.rs:142]: Hmm I agree, but isn't that an unstable feature (https://doc.rust-lang.org/unstable-book/language-features/trait-alias.html)?

**Comment by etienne-k** [comment]
Thank you for the review, @sudormrfbin ! I extended the book a bit to include some configuration documentation (000a742 & 6699c05).
I corrected/changed the code according to the review as well; I do have one open remark regarding trait aliases, see above.

File: helix-term/src/ui/statusline.rs
Comment (etienne-k) [reply helix-term/src/ui/statusline.rs:65]: Hold on, we were too fast here... This breaks the statusline style. Reverted this change in 444fd95.

File: book/src/configuration.md
Comment (sudormrfbin) [inline book/src/configuration.md:73]: Since we only show the encoding if it's not utf-8 it should be explicitly mentioned here.

File: helix-term/src/ui/statusline.rs
Comment (sudormrfbin) [reply helix-term/src/ui/statusline.rs:142]: Ah in that case let's leave it as it is 👍🏾

File: book/src/configuration.md
Comment (etienne-k) [reply book/src/configuration.md]: Agree. Changed in 5621494.

**Comment by etienne-k** [comment]
Alright, I rebased the branch and reintegrated the last few bits of #2676 in 8d28f95. All looks good and clippy is happy! :)

File: helix-term/src/ui/statusline.rs
Comment (pickfire) [inline helix-term/src/ui/statusline.rs:187]: Should we keep all these functions within the same struct? They don't even need `Self`, would it be better if they are splitted out into functions like typed commands?

Comment (sudormrfbin) [reply helix-term/src/ui/statusline.rs:187]: Agree, that would also be in line with gutter components which are just functions in `gutter.rs`.

Comment (etienne-k) [reply helix-term/src/ui/statusline.rs:187]: Do you suggest completely getting rid of the `StatusLine` struct? Because none of the functions require `Self`.

I initially designed it that way assuming I would need a context in `Self`, but I figured it'd be better passing the context as a parameter.

Comment (etienne-k) [reply helix-term/src/ui/statusline.rs]: Anyway, since the struct wasn't really needed, I refactored it away in fae2348.

**Comment by etienne-k** [comment]
As a friendly side note: I don't mind refactoring more of this after the PR was merged, even if it's just cosmetic code changes! :)

Not that I mind doing so now, but we're starting to block some other features which build upon this, and there's always the risk that we have to rebase onto more changes like #2676.

File: helix-term/src/ui/statusline.rs
Comment (Mathspy) [reply helix-term/src/ui/statusline.rs:173]: Also sorry for https://github.com/helix-editor/helix/pull/2676 and the painful rebase 😓

Comment (etienne-k) [reply helix-term/src/ui/statusline.rs:173]: I've been thinking about your comment, but I'm not entirely convinced it'd be a good idea to treat the mode and colored mode as separate components, for two reasons:
- I can't think of a realistic use case where someone would want _both_ a colored mode _and_ a monochrome mode.
- When designing the code, I specifically included an optional `Style` in each write callback, for exactly this reason: Allowing render functions to - under certain circumstances - define a `Style`.
  - In fact, I was planning on implementing exactly that feature as soon as this PR would be merged! :)

Comment (Mathspy) [reply helix-term/src/ui/statusline.rs]: > - I can't think of a realistic use case where someone would want _both_ a colored mode _and_ a monochrome mode.

Yep fair enough!

> - When designing the code, I specifically included an optional `Style` in each write callback, for exactly this reason: Allowing render functions to - under certain circumstances - define a `Style`.

That's pretty awesome ❤️

File: helix-term/src/ui/statusline.rs
Comment (yuanyelele) [reply helix-term/src/ui/statusline.rs:287]: I think `doc.language_id` is broken, because it almost always returns from the last `or_else`, even for rust source.

Comment (Mathspy) [reply helix-term/src/ui/statusline.rs:287]: I handled this in https://github.com/helix-editor/helix/pull/2998 so that you don't have to worry about it in this PR

File: helix-term/src/ui/statusline.rs
Comment (archseer) [inline helix-term/src/ui/statusline.rs:188]: There seems to be additional spacing between `NOR` and the spinner now (and also on the right side).

I think the spinner should have no padding. We can also add a spacer element that simply renders an empty space so that users can space out the components as they wish

File: helix-term/src/ui/statusline.rs
Comment (archseer) [inline helix-term/src/ui/statusline.rs:157]: Keep the spacing here though since mode sets a background color

**Review by archseer — CHANGES_REQUESTED** [review]
Tested and looks fine, I just have one nit about the styling not matching current master

File: helix-term/src/ui/statusline.rs
Comment (sbromberger) [reply helix-term/src/ui/statusline.rs:188]: You'll also be able to add spacing within a custom spinner with #2793.

Comment (etienne-k) [reply helix-term/src/ui/statusline.rs]: Oh you're right, seems like I did change the padding compared to before this PR.
I changed it back in dba88ba.

With spinner padding:
image

Without spinner padding (master):
image

I must've introduced the padding intuitively, as it made more sense to me.

File: helix-term/src/ui/statusline.rs
Comment (etienne-k) [reply helix-term/src/ui/statusline.rs:82]: Hm, do you mean something like introducing a function like the following and calling it in `render()`?
```rust
fn render_area(element_ids: &Vec<StatusLineElement>, write: F,  x: u16, y: y16)
    where F: Fn(&mut RenderContext, String, Option<Style>) + Copy
```

**Review by archseer — APPROVED** [review]
Tested, works great! Thanks for working on this :)

**Pull request merged**
